### PR TITLE
[stable/kong] use volumes for runtime data

### PR DESCRIPTION
**Pull request by rainest**
#### Is this a new chart
No.

#### What this PR does / why we need it:
These changes allow enforcing Kubernetes security policies that restrict writes to the root filesystem by mounting emptyDir volumes over directories that Kong needs to write to at runtime.

#### Special notes for your reviewer:
At present, Kong Enterprise has a bug where it still attempts to write to /usr/local/kong, so this does not allow using it with a read-only root. It should work once that bug is addressed. Older versions (<=0.36-2) furthermore have issues with non-standard prefixes. 1.3+ should not encounter issues with the non-standard prefix, and I did not encounter any testing with these changes in place while still allowing root writes.

Core should work with the change and a read-only root as is.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

File: stable/kong/templates/_helpers.tpl [hunk 1]
```diff
@@ -177,6 +181,10 @@ the extra slash.
 {{- end -}}
 
 {{- define "kong.volumeMounts" -}}
+- name: {{ template "kong.fullname" . }}-prefix-dir
+  mountPath: /kong_prefix/
+- name: {{ template "kong.fullname" . }}-tmp
+  mountPath: /tmp
```
Comment (hbagdi) [inline stable/kong/templates/_helpers.tpl:211]: Why is this needed? :thinking:

File: stable/kong/values.yaml [hunk 2]
```diff
@@ -283,6 +283,7 @@ env:
   admin_error_log: /dev/stderr
   admin_gui_error_log: /dev/stderr
   portal_api_error_log: /dev/stderr
+  prefix: /kong_prefix/
```
Comment (hbagdi) [inline stable/kong/values.yaml:286]: This is kind of weird.
Software packages don't usually create a directory under `/`.
The practice I've seen most widely used is to create such a working directory in `/opt`.

I'm not sure but there might be some security policies to stop processes from writing to `/`.

Thoughts?

(Of course, Kong should not log to that but under `/var/log` but that's a bigger conversation for another time.)

**Review by hbagdi — COMMENTED** [review]
Looks good mostly but I've a couple questions.

File: stable/kong/templates/_helpers.tpl (hunk 1 above)
Comment (rainest) [reply stable/kong/templates/_helpers.tpl:211]: NGINX creates some files in /tmp briefly during startup, and I don't believe we have control to have it use another directory.

File: stable/kong/values.yaml (hunk 2 above)
Comment (rainest) [reply stable/kong/values.yaml:286]: We live in container land and traditional Unix hierarchy no longer matters :D

The choice was essentially arbitrary, but based on the existing mounts at `/kong` and `/kong_dbless` at https://github.com/helm/charts/blob/master/stable/kong/templates/_helpers.tpl#L179-L185

As far as I know, the only Kubernetes-level restrictions are on allowing root writes (at the mount level) and what types of volumes can be written to. The second part probably needs further review: the existing PSP template [does not allow emptyDir writes](https://github.com/helm/charts/blob/master/stable/kong/templates/psp.yaml#L23-L25), which should break this. However, I didn't encounter issues when enabling the PSP, which is curious.

Standard FS permissions apply within the container, but anything at the root gets its own set of permissions (i.e. there shouldn't ever be any issue traversing to it for lack of execute on an intermediate directory), and emptyDirs get mounted with all access to everyone:

```
ls -ld /kong_prefix/
drwxrwxrwx 11 root root 4096 Dec 10 19:38 /kong_prefix/
[kong@scared-koala-kong-d7f8748f5-ml9jp /]$ env | grep PATH
KONG_LUA_PACKAGE_PATH=/opt/?.lua;;
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```
Part of the goal of the read-only root configuration is to avoid having writeable locations in the PATH, to guard against an attacker writing and then invoking malicious code or binaries. Our usage of `/opt` in the Lua path should preclude placing the prefix under there as such, though we separately need to make those ConfigMap mounts read-only also for that to work.

**Comment by rainest** [comment]
Current WIP status: https://github.com/Kong/docker-kong/issues/313 breaks using non-standard prefixes if the container runs as root, but then invokes Kong with the entrypoint script. Adding securityContexts throughout the templates to ensure Kubernetes runs Kong as not root, which we should do regardless.

**Comment by rainest** [comment]
Closing and reopening elsewhere. This ended up requiring many more changes than initially expected, to the point that I think it warrants rewriting the PR.
